### PR TITLE
feat: store fees config for contractmethod

### PR DESCRIPTION
**Pull request by epsjunior**
# Store Fees Configuration Per Contract+Method Combination

## Summary
Enhances the GenLayer wallet snap to store fees configuration per contract+method combination instead of just per contract address. This improvement allows different methods within the same contract to have separate fee configurations, providing more granular control and better user experience. The implementation includes GenLayer transaction parsing, composite storage key generation, and comprehensive test coverage.

### 🚀 New Features
- **Per-Method Fee Storage**: Fees configuration now stored separately for each contract method combination
- **GenLayer Transaction Parsing**: Proper extraction of contract address and method name from GenLayer transactions
- **Composite Storage Keys**: Storage keys in format `contractAddress_methodName` (e.g., `0x123...abc_transfer`)
- **Method Name Extraction**: Real method names like `transfer`, `approve`, `mint` instead of hex selectors
- **Fallback Handling**: Graceful defaults (`default_unknown`) when parsing fails

### 🔧 Core Implementation
- **Contract Address Extraction**: Now uses `parsed.args[1]` instead of `transaction.to` for actual contract address
- **GenLayer Integration**: Uses `genlayer-js` and `ethers` for proper transaction decoding
- **Storage Key Generation**: Composite keys combining normalized contract address and method name
- **Error Resilience**: Robust error handling with fallbacks to prevent snap failures

### 🏗️ Architecture Updates
- **Transaction Parsing Service**: Centralized transaction handling in dedicated folder
- **State Management Enhancement**: Updated handlers to use composite storage keys
- **Interface Consistency**: All user interactions now work with per-method configurations
- **Clean Codebase**: Removed debug logging for production-ready code

### 📁 File Structure
```
packages/snap/src/
├── transactions/           # New folder (renamed from utils)
│   ├── transaction.ts      # Core transaction parsing logic
│   └── transaction.test.ts # Comprehensive test coverage
├── index.tsx              # Updated handlers for composite keys
├── index.test.tsx         # Updated test expectations
├── libs/
│   └── StateManager.tsx   # Clean state management (no logs)
└── components/
    ├── TransactionConfig.tsx    # Clean component rendering
    └── AdvancedOptionsForm.tsx  # Clean form handling
```

### 🔍 Transaction Parsing Logic
```typescript
// Before: Only contract address
transaction.to → "0x123...abc"

// After: Contract + Method combination
parseGenLayerTransaction(data) → {
  contractAddress: "0x123...abc",  // From parsed.args[1]
  methodName: "transfer"           // From decoded calldata
}

// Storage Key Generation
generateStorageKey("0x123...abc", "transfer") → "0x123...abc_transfer"
```

### 🎯 GenLayer Integration Details
- **Consensus Contract Parsing**: Uses `chains.localnet.consensusMainContract.abi`
- **RLP Decoding**: Proper decoding of `parsed.args[4]` transaction data
- **Calldata Extraction**: Uses `abi.calldata.decode()` for method name extraction
- **Type Safety**: Full TypeScript integration with proper error handling

### 🧪 Testing Coverage
- **Transaction Parsing Tests** (`transactions/transaction.test.ts`):
  - Valid GenLayer transaction parsing
  - Contract address extraction from `args[1]`
  - Method name extraction from decoded calldata
  - Fallback behavior for parsing failures
  - Edge cases with missing data

- **Handler Tests** (`index.test.tsx`):
  - Updated mock expectations for composite keys
  - Storage operations with contract+method keys
  - Form submission and retrieval with new key format

- **Mock Infrastructure**:
  - Complete mocking of `genlayer-js` and `ethers` dependencies
  - Realistic test scenarios with actual transaction data
  - Error condition testing for robustness

### 🔄 State Management Flow
```typescript
// Previous Flow
onTransaction: transaction.to → "0x123...abc"
onUserInput: StateManager.get("0x123...abc") → shared config

// New Flow  
onTransaction: getTransactionStorageKey(transaction) → "0x123...abc_transfer"
onUserInput: StateManager.get("0x123...abc_transfer") → method-specific config
```

### 📋 Storage Key Examples
```
Contract: 0x123...abc
├── transfer method → "0x123...abc_transfer"
├── approve method  → "0x123...abc_approve" 
├── mint method     → "0x123...abc_mint"
└── unknown method  → "0x123...abc_unknown"

Contract: 0x456...def  
├── burn method     → "0x456...def_burn"
└── pause method    → "0x456...def_pause"
```

### 🛠️ Implementation Benefits
- **Granular Control**: Each method can have different fee configurations
- **Better UX**: Previous fee settings remembered per method, not just contract
- **Scalability**: No conflicts between methods in the same contract
- **Maintainability**: Clean separation of concerns with dedicated transaction handling

### 🔧 Error Handling
- **Parsing Failures**: Graceful fallback to `default_unknown` storage key
- **Missing Data**: Handles transactions without proper GenLayer structure
- **Type Validation**: Ensures contract addresses and method names are valid strings
- **No Breaking Changes**: Snap continues to function even with invalid data

### 🎨 Code Quality Improvements
- **No Debug Logs**: Clean production code without console.log statements
- **Professional Naming**: Renamed `utils` folder to `transactions` for better semantics
- **Comprehensive Comments**: Clear JSDoc documentation for all functions
- **TypeScript Integration**: Full type safety with proper interfaces

### 🔄 Migration Impact
- **Backward Compatibility**: New installations will use per-method storage
- **Existing Data**: Previous contract-only configurations remain accessible
- **No Data Loss**: Existing fee configurations continue to work
- **Progressive Enhancement**: New transactions get per-method benefits

### ✨ Technical Implementation
- **Pure Functions**: All transaction parsing functions are side-effect free
- **Immutable Operations**: No mutation of input parameters
- **Error Boundaries**: Try-catch blocks prevent snap crashes
- **Memory Efficient**: Minimal overhead for storage key generation

### 🎯 User Experience
- **Transparent**: Users don't need to understand the technical changes
- **Improved**: Fee configurations are more precise and persistent
- **Reliable**: Robust error handling prevents configuration loss
- **Intuitive**: Same UI with enhanced backend functionality

### 🔗 Dependencies
- **genlayer-js**: For GenLayer-specific transaction parsing
- **ethers**: For RLP decoding and byte manipulation
- **@metamask/snaps-sdk**: Existing snap framework integration
- **No New Dependencies**: Uses existing project dependencies

---

**Testing**: ✅ All tests pass with comprehensive coverage  
**Type Safety**: ✅ Full TypeScript integration with proper error handling  
**Documentation**: ✅ Clear JSDoc comments and inline documentation  
**Architecture**: ✅ Clean separation with dedicated transaction handling  
**User Experience**: ✅ Seamless upgrade with enhanced functionality  
**Production Ready**: ✅ No debug logs, clean professional codebase  
**GenLayer Integration**: ✅ Proper transaction parsing using official libraries

## Summary by CodeRabbit

* **New Features**
  * Introduced utilities for parsing GenLayer transaction data and generating transaction-specific storage keys.
* **Bug Fixes**
  * State retrieval now returns null instead of undefined when a key is missing.
* **Tests**
  * Added comprehensive tests for transaction parsing and storage key generation utilities.
  * Updated existing tests to use the new transaction-specific storage keys.
* **Chores**
  * Updated dependencies and manifest checksums to reflect recent changes.

### CODE REVIEW

File: packages/snap/src/transactions/transaction.ts [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+
+/**
+ * Extracts contract address and method name from GenLayer transaction data
+ * @param data - The transaction data (hex string)
+ * @returns Object with contractAddress and methodName, or defaults if extraction fails
+ */
+export function parseGenLayerTransaction(data: string): { contractAddress: string; methodName: string } {
+  try {
+    const contractInterface = new Interface(chains.localnet.consensusMainContract?.abi as any);
+    const parsed = contractInterface.parseTransaction({ data });
```
[the remark this answers is not in the document]
Comment (epsjunior) [reply packages/snap/src/transactions/transaction.ts:16]: we already have a fallback if decoding/parsing fails. (lines 36-43)

File: packages/snap/src/transactions/transaction.test.ts [hunk 2]
```diff
@@ -0,0 +1,334 @@
...
+    it('should generate composite key with contract address and method name', () => {
+      const contractAddress = '0x1234567890123456789012345678901234567890';
+      const methodName = 'transfer';
+      const result = generateStorageKey(contractAddress, methodName);
+      expect(result).toBe(
+        '0x1234567890123456789012345678901234567890_transfer',
+      );
+    });
+
+    it('should handle uppercase contract address by converting to lowercase', () => {
```
Comment (cristiam86) [inline packages/snap/src/transactions/transaction.test.ts:212]: @epsjunior this test talks about uppercase contract address but the address is all numbers

**Review by cristiam86 — CHANGES_REQUESTED** [review]
@epsjunior please fix test case

File: packages/snap/src/transactions/transaction.test.ts (hunk 2 above)
Comment (epsjunior) [reply packages/snap/src/transactions/transaction.test.ts:212]: @cristiam86  the original contract address was 0X1234567890123456789012345678901234567890 — all digits, but with an uppercase "X" in the 0X prefix. It was normalized to lowercase, becoming 0x1234567890123456789012345678901234567890.

So although the address itself is numeric, the test is still valid — it verifies that even a single uppercase character like the "X" is properly converted to lowercase.

Comment (epsjunior) [reply packages/snap/src/transactions/transaction.test.ts:212]: changed for this address: 0xc361Fc33b99F88612257ac8cC2d852A5CEe0E217

**Review by cristiam86 — APPROVED** [review]
